flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/io/CompressedHeaderlessChannelReaderInputView.java

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.io;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.disk.iomanager.AbstractChannelReaderInputView;
+import org.apache.flink.runtime.io.disk.iomanager.BufferFileReader;
+import org.apache.flink.runtime.io.disk.iomanager.FileIOChannel;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.io.disk.iomanager.RequestDoneCallback;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.table.runtime.compression.BlockCompressionFactory;
+import org.apache.flink.table.runtime.compression.BlockCompressor;
+import org.apache.flink.table.runtime.compression.BlockDecompressor;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ *
+ */
+public class CompressedHeaderlessChannelReaderInputView
+		extends AbstractChannelReaderInputView
+		implements RequestDoneCallback<Buffer>, BufferRecycler {
+
+	private final BlockDecompressor decompressor;
+	private final BufferFileReader reader;
+	private final MemorySegment uncompressedBuffer;
+	private final AtomicReference<IOException> cause;
+
+	private final LinkedBlockingQueue<Buffer> retBuffers = new LinkedBlockingQueue<>();
+
+	private int numBlocksRemaining;
+	private int currentSegmentLimit;
+
+	public CompressedHeaderlessChannelReaderInputView(
+			FileIOChannel.ID id,
+			IOManager ioManager,
+			BlockCompressionFactory compressionCodecFactory,
+			int compressionBlockSize,
+			int numBlocks) throws IOException {
+		super(0);
+		this.numBlocksRemaining = numBlocks;
+		this.reader = ioManager.createBufferFileReader(id, this);
+		uncompressedBuffer = MemorySegmentFactory.wrap(new byte[compressionBlockSize]);
+		decompressor = compressionCodecFactory.getDecompressor();
+		cause = new AtomicReference<>();
+
+		BlockCompressor compressor = compressionCodecFactory.getCompressor();
+		for (int i = 0; i < 2; i++) {
+			MemorySegment segment = MemorySegmentFactory.wrap(new byte[compressor.getMaxCompressedSize(
+					compressionBlockSize)]);
+			reader.readInto(new NetworkBuffer(segment, this));
+		}
+	}
+
+	@Override
+	protected MemorySegment nextSegment(MemorySegment current) throws IOException {
+		if (cause.get() != null) {
+			throw cause.get();
+		}
+
+		// check for end-of-stream
+		if (this.numBlocksRemaining <= 0) {
+			this.reader.close();
+			throw new EOFException();
+		}
+
+		try {
+			Buffer buffer;
+			while ((buffer = retBuffers.poll(1, TimeUnit.SECONDS)) == null) {
+				if (cause.get() != null) {
```

Comment:
every loop should check closed too?
same to AsynchronousBlockReader.getNextReturnedBlock